src/System.IO.Pipes/src/System/IO/Pipes/PipeStream.Windows.cs

```diff
@@ -847,71 +521,13 @@ internal static Interop.mincore.SECURITY_ATTRIBUTES GetSecAttrs(HandleInheritabi
             return secAttrs;
         }
 
-        // When doing IO asynchronously (i.e., _isAsync==true), this callback is 
-        // called by a free thread in the threadpool when the IO operation 
-        // completes.  
-        [SecurityCritical]
-        unsafe private static void AsyncPSCallback(uint errorCode, uint numBytes, NativeOverlapped* pOverlapped)
+        internal void UpdateMessageCompletion(bool completion)
         {
-            // Extract async result from overlapped 
-            PipeStreamAsyncResult asyncResult = (PipeStreamAsyncResult)ThreadPoolBoundHandle.GetNativeOverlappedState(pOverlapped);
-            asyncResult._numBytes = (int)numBytes;
-
-            // Allow async read to finish
-            if (!asyncResult._isWrite)
-            {
-                if (errorCode == Interop.mincore.Errors.ERROR_BROKEN_PIPE ||
-                    errorCode == Interop.mincore.Errors.ERROR_PIPE_NOT_CONNECTED ||
-                    errorCode == Interop.mincore.Errors.ERROR_NO_DATA)
-                {
-                    errorCode = 0;
-                    numBytes = 0;
-                }
-            }
-
-            // For message type buffer.
-            if (errorCode == Interop.mincore.Errors.ERROR_MORE_DATA)
-            {
-                errorCode = 0;
-                asyncResult._isMessageComplete = false;
-            }
-            else
-            {
-                asyncResult._isMessageComplete = true;
-            }
-
-            asyncResult._errorCode = (int)errorCode;
-
-            // Call the user-provided callback.  It can and often should
-            // call EndRead or EndWrite.  There's no reason to use an async 
-            // delegate here - we're already on a threadpool thread.  
-            // IAsyncResult's completedSynchronously property must return
-            // false here, saying the user callback was called on another thread.
-            asyncResult._completedSynchronously = false;
-            asyncResult._isComplete = true;
-
-            // The OS does not signal this event.  We must do it ourselves.
-            ManualResetEvent wh = asyncResult._waitHandle;
-            if (wh != null)
-            {
-                Debug.Assert(!wh.GetSafeWaitHandle().IsClosed, "ManualResetEvent already closed!");
-                bool r = wh.Set();
-                Debug.Assert(r, "ManualResetEvent::Set failed!");
-                if (!r)
-                {
-                    throw Win32Marshal.GetExceptionForLastWin32Error();
-                }
-            }
-
-            AsyncCallback callback = asyncResult._userCallback;
-            if (callback != null)
-            {
-                callback(asyncResult);
-            }
+            // Set message complete to true because the pipe is broken as well.
+            // Need this to signal to readers to stop reading.
+            _isMessageComplete = (_state == PipeState.Broken || completion);
```

Comment:
Very minor, but consider swapping the order of these to be ```completion || (_state == PipeState.Broken)``` so that we can avoid reading the field if completion is true.